src/lib/source-handlers/bitbucket-server/list-projects.ts

```diff
@@ -0,0 +1,87 @@
+import * as needle from 'needle';
+import * as debugLib from 'debug';
+import Bottleneck from 'bottleneck';
+import { BitbucketServerProjectData } from './types';
+import { getBitbucketServerToken } from './get-bitbucket-server-token';
+
+const debug = debugLib('snyk:bitbucket-server');
+
+const limiter = new Bottleneck({
+  reservoir: 1000, // initial value
+  reservoirRefreshAmount: 1000,
+  reservoirRefreshInterval: 3600 * 1000,
+  maxConcurrent: 1,
+  minTime: 1000,
+});
+
+limiter.on('failed', async (error, jobInfo) => {
+  const id = jobInfo.options.id;
+  debug(`Job ${id} failed: ${error}`);
+  if (jobInfo.retryCount === 0) {
+    // Here we only retry once
+    debug(`Retrying job ${id} in 25ms!`);
+    return 25;
+  }
+});
+
+export const fetchAllProjects = async (
+  url: string,
+  token: string,
+): Promise<unknown[]> => {
+  let isLastPage = false;
+  let values: unknown[] = [];
+  let pageCount = 1;
+  let start = 0;
+  const limit = 100;
+  while (!isLastPage) {
+    debug(`Fetching page ${pageCount} for projects\n`);
+    const response = await limiter.schedule(() =>
+      needle(
+        'get',
+        `${url}/rest/api/1.0/projects?start=${start}&limit=${limit}`,
+        {
+          headers: { Authorization: 'Bearer ' + token },
+        },
+      ),
+    );
+    if (response.statusCode != 200) {
+      if (response.statusCode == 429) {
+        debug(
+          `Failed to fetch page: ${url}\n, Response Status: ${response.body}\nToo many requests \nWaiting for 3 minutes before resuming`,
+        );
+        await sleepNow(180000);
+        isLastPage = false;
+      } else {
+        debug(
+          `Failed to fetch page: ${url}\n, Response Status: ${response.statusCode}\nResponse Status Text: ${response.body} `,
+        );
+      }
+    }
+    const apiResponse = response.body['values'];
+    values = values.concat(apiResponse);
```

Comment:
what is values? Can we rename it to mean something in the context of this function and not related to BBS

Reply:
done